File: 238.ts
// 238. Product of Array Except Self
// https://leetcode.com/problems/product-of-array-except-self/

function productExceptSelf1(nums: number[]): number[] {
  return nums.map((num, index) => {
    let result = 1;

    for (let j = 0; j < nums.length; j++) {
      if (index !== j) {
        result *= nums[j];
      }
    }

    return result;
  });
}

function productExceptSelf2(nums: number[]): number[] {
  const memo: number[] = [];

  return nums.map((num, index) => {
    let result = 1;

    if (index - 1 >= 0 && memo[index - 1]) {
      result *= memo[index - 1];

      for (let i = index + 1; i < nums.length; i++) {
        result *= nums[i];
      }
    } else {
      let valueToMemo = 1;
      for (let i = 0; i <= index - 1; i++) {
        valueToMemo *= nums[i];
      }

      memo[index - 1] = valueToMemo;

      result *= memo[index - 1];

      for (let i = index + 1; i < nums.length; i++) {
        result *= nums[i];
      }
    }

    return result;
  });
}

File: LeetCode - Algorithm/Array - String/238.java
// 238. Product of Array Except Self
// https://leetcode.com/problems/product-of-array-except-self/

import java.util.Arrays;

class Solution {
    // time complexity: O(2n) = O(n)
    public int[] productExceptSelf(int[] nums) {
        int zeroCount = 0;
        int zeroIndex = -1;

        int allValuesProducted = 1;

        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];

            if (num == 0 && zeroCount == 0) {
                zeroCount += 1;
                zeroIndex = i;
            } else if (num == 0 && zeroCount == 1) {
                int[] zeros = new int[nums.length];
                Arrays.fill(zeros, 0);

                return zeros;
            } else {
                allValuesProducted *= num;
            }
        }

        int[] result = new int[nums.length];
        Arrays.fill(result, 0);

        if (zeroCount == 1)
            result[zeroIndex] = allValuesProducted;
        else
            for (int i = 0; i < nums.length; i++) {
                result[i] = allValuesProducted / nums[i];
            }

        return result;
    }
}